### ISSUE TITLE
math for delta exponent (rotation) derivation

### ISSUE THREAD

**Issue by rdgao** [fooof-tools/fooof#128]
@TomDonoghue can you assign this to me? Just so I have a paper trail to come back to after cosyne, otherwise I might forget.

**Comment by TomDonoghue** [comment]
Done! (Though I think you should have also been able to assign yourself..?). 

For tracking: this issue is related to adding notes and a description of the math that was merged in #127 and discussed in #109.

**Comment by TomDonoghue** [comment]
Throwing a couple things here (we've been playing with rotations, and just noting some properties).

Broadly, we ran into some issues related to using a lot of rotations, in the sense that if you start with a spectra, A, and create rotated version of the spectra B & C, the relationship between B & C is not necessarily super obvious based on what you applied to A (but if one is rotating a bunch of spectra, it might be the relationship between B & C that is of interest). 

Here's what it seems to do, for some cases. 

1) Rotate A, with the same delta-exponent for B & C, but at different rotation frequencies: 
The resulting power spectra have an offset difference, that is predictable at the difference of the delta offsets of each rotation. 

In code:
```
# Rotate a spectrum, the same amount, at two different rotation frequencies
ps1 = rotate_spectrum(fs, ps, delta_exp, f_rot1)
ps2 = rotate_spectrum(fs, ps, delta_exp, f_rot2)
# The off difference between ps2 and ps1 is the difference of the delta offsets from each rotation. 
off_diff = compute_rotation_offset(delta_exp, f_rot1) - compute_rotation_offset(delta_exp, f_rot2)
```

2) Rotate A, with different delta-exponents for B & C, at the same rotation frequencies: 
The resulting power spectra have an exponent difference, around the specified rotation frequency, that is the difference of the delta exponents. 

In code:
```
ps1 = rotate_spectrum(fs, ps, delta_exp1, f_rot)
ps2 = rotate_spectrum(fs, ps, delta_exp2, f_rot)
exp_diff = delta_exp1-delta_exp2
```

3)  Rotate A, with different delta-exponents for B & C, and different rotation frequencies: 
The resulting power spectra have an exponent difference, and an offset difference, both which are predictable. However, they also have a rotation frequency between them, that I don't know how to predict. 

In code:
```
ps1 = rotate_spectrum(fs, ps, delta_exp1, f_rot1)
ps2 = rotate_spectrum(fs, ps, delta_exp2, f_rot2)
exp_diff = delta_exp1-delta_exp2
off_diff = compute_rotation_offset(delta_exp1, f_rot1) - compute_rotation_offset(delta_exp2, f_rot2)
new_f_rot = ???
```

Basically just copying this all in because maybe we want to include it in a tutorial or something. Once I starting thinking this through, everything basically makes sense and is relatively straight forward, but it wasn't immediately obvious when we started doing rotations that these things would happen. 

So the only thing I don't know is how to predict f_rotations between two power spectra given that you've rotated them. I presume there is a solution to this given the equation we use to rotate, if you happen to know what it is @rdgao 

cc: @ArcadeShrimp - this is some details on the rotation things we've been talking through the last couple days.

**Comment by rdgao** [comment]
here's the math for the new one-liner rotate_spectrum:

rotation in loglog is equivalent to multiplying the original spectrum with a 1/f-powerlaw shaped weighting mask that is 
1) unity at the rotation frequency of choice, (i.e., mask[f_rot] = 1), and 
2) has an exponent of delta_exp, which down/up-weights power in lower/higher frequencies for counterclockwise rotation, or vice versa for clockwise rotation. In math:

if psd_new = psd_orig * mask, 
then psd_new is assumed to have the form: 
psd_new ~ 1/f**(exp_orig+delta_exp), and psd_new[f_rot] = psd_orig[f_rot]. 

Therefore,
mask = A/f**delta_exp, and A = f_rot**delta_exp, 
substituting A, we have
mask = (f_rot/f)**delta_exp, or (f/f_rot)**-delta_exp.

In code: `mask = (np.abs(freqs) / f_rotation)**-delta_exponent`, where np.abs() ensures the same transformation to both positive and negative frequencies.

**Comment by rdgao** [comment]
for situation 3 above:
B = A*(f_rotB/f)^delta_expB, C = A*(f_rotC/f)^delta_expC. 
Therefore, what you want is f_rot_BC where that is the frequency such that B=C. So we plug everything back into the equation, finding f s.t. B[f] = C[f]:

(f_rotB/f)^delta_expB = (f_rotC/f)^delta_expC (dropping A on both sides),
f_rotB^delta_expB * f^-delta_expB = f_rotC^delta_expC * f^-delta_expC, collect f on the same side: 
(f_rotB^delta_expB)/(f_rotC^delta_expC) = f^-(delta_expC+delta_expB), so
f_rotBC = [(f_rotB^delta_expB)/(f_rotC^delta_expC)]^-1/(delta_expC+delta_expB), or 
[(f_rotC^delta_expC)/(f_rotB^delta_expB)]^1/(delta_expC+delta_expB)

see if that works?

**Comment by TomDonoghue** [comment]
Tagging @rdgao back in. 

Thanks for the notes on the transforms here - I have finally added some better annotation into the codebase, and for cases 1 & 2 that are described above I'm dropping these into the example, so people can see how this plays out. 

Current problem: for case 3, the proposed solution above does not seem to calculate the right thing. I'm not quite sure where the math is going wrong - so if you have a chance to look into it that'd be great. 

Code for the proposal:
```
def calc_f_rot(f_rot_b, delta_b, f_rot_c, delta_c):
    
    return (((f_rot_c**delta_c) / (f_rot_b**delta_b))) ** (1/(delta_c+delta_b))
```

Example I'm using to test:

```
import matplotlib.pyplot as plt

from fooof.sim import gen_power_spectrum
from fooof.sim.transform import rotate_spectrum
from fooof.plts import plot_spectra

delta_exp1 = 0.5
delta_exp2 = 0.75
f_rot1 = 2
f_rot2 = 5

fs, ps = gen_power_spectrum([3, 50], [0, 2], [10, 0.25, 0.5], nlv=0)

ps1 = rotate_spectrum(fs, ps, delta_exp1, f_rot1)
ps2 = rotate_spectrum(fs, ps, delta_exp2, f_rot2)

fr_1 = calc_f_rot(f_rot2, delta_exp2, f_rot1, delta_exp1)
fr_2 = calc_f_rot(f_rot1, delta_exp1, f_rot2, delta_exp2)

print(fr_1, fr_2)

plot_spectra(fs, [ps1, ps2], True, True)
plt.axvline(fr_1)
plt.axvline(fr_2)
```

If you run this code (should be self-contained) the outputs get that `fr_1` and `fr_2` are different (which I believe they should be the same), and neither of seem appear to capture the rotation frequency between B & C. 

I'm also unclear on if the outputs `fr_1` & `fr_2` are, or should be, in log space or linear space.

**Comment by TomDonoghue** [comment]
@rdgao : if you get a chance to look at this over the next ~week, I'd like to try and fold any notes from it into the 1.0 release :)

**Comment by rdgao** [comment]
oh shit okay I will figure this out in the next few days
rg.

**Comment by rdgao** [comment]
lol. amateur hour, @rdgao. I subbed something wrong in the derivation above, it should be:
```
def calc_f_rot(f_rot_b, delta_b, f_rot_c, delta_c):
    
    return (((f_rot_c**delta_c) / (f_rot_b**delta_b))) ** (1/(delta_c-delta_b))
```
where the `1/(delta_c + delta_b)` should be `1/(delta_c - delta_b)`.

rest of the example will work as expected, but axvlines were drawn on the wrong place because frequencies were not logged. Added that to the snippet:

```
import matplotlib.pyplot as plt
from fooof.sim import gen_power_spectrum
from fooof.sim.transform import rotate_spectrum
from fooof.plts import plot_spectra
from numpy import log10

delta_exp1 = 0.5
delta_exp2 = 0.75
f_rot1 = 2
f_rot2 = 5

fs, ps = gen_power_spectrum([3, 50], [0, 2], [10, 0.25, 0.5], nlv=0)

ps1 = rotate_spectrum(fs, ps, delta_exp1, f_rot1)
ps2 = rotate_spectrum(fs, ps, delta_exp2, f_rot2)

fr_1 = calc_f_rot(f_rot2, delta_exp2, f_rot1, delta_exp1)
fr_2 = calc_f_rot(f_rot1, delta_exp1, f_rot2, delta_exp2)

print(fr_1, fr_2)

plot_spectra(fs, [ps1, ps2], True, True)
plt.axvline(log10(fr_1))
plt.axvline(log10(fr_2))
```

**Comment by TomDonoghue** [comment]
@rdgao awesome - thanks a lot! Glad it was a relatively easy fix. 

Updates for this issue:
- added `compute_rotation_frequency`, using fixed derivation from above
  - Direct link: https://github.com/fooof-tools/fooof/blob/master/fooof/sim/transform.py#L187
- added a description of relations between spectra to the examples
  - Direct link: https://github.com/fooof-tools/fooof/blob/master/examples/sims/plot_transforms.py#L125

Unless anything looks funky here, or there are any other related suggestions, I think we're good on this issue!

**Comment by TomDonoghue** [comment]
This seems to all be covered now, so I'm going to close this!
